(WIP) Add draggable/droppable components and redefine interfaces resolve #16

yihyun-kim1 opened the pull request:
기존에 정해져있던 코스를 변경할 수 있도록 드래그 앤 드롭 기능을 포함하는 컴포넌트를 구성했습니다. (진행중)

- 원래는 공통 컴포넌트를 제작하는 과정 중에 있었으나 해당 컴포넌트를 사용하는 타 페이지는 없으며
- 기존의 데이터와 버튼을 통한 API 호출, 즉각적인 변화를 반영해야 하는 데이터 특성상 해당 컴포넌트를 사용하는 flow 전체를 구현하는 PR로써 동작할 것 같습니다.

*Type 관련해서는 여러 번의 시행착오를 겪은 만큼 더 고민해서 올려보겠습니다. 빌드 에러 빠른 시일 내로 해결해서 커밋하겠습니다!

0707 update)
- Object type을 사용해 dnd 기능을 완성했습니다. 추가/삭제 시에도 정상적으로 전체 인덱스가 관리되도록 globalIndex를 활용해 구현했고, react 18 버전에서 strictMode로 인한 제약을 해결하기 위한 방안으로 StrictModeDroppable 컴포넌트를 사용했습니다.
- WIP) 동일한 타입 내에서 아이템의 개수가 늘어나거나 감소할 경우 "N차" 라벨 부착 관련 기능은 보완이 필요합니다. 
- 아이템이 클릭되었을 때 바텀시트가 닫히지 않는 이슈를 해결해야 합니다.
*바텀시트는 비밀번호 입력 시에도 사용하지만, 내부적인 로직이나 UI 상으로 재사용하기 어려워보여 해당 페이지 전용으로 사용해도 될 지 의견 부탁드립니다!

yihyun-kim1 committed:
feat: Add draggable/droppable components and redefine interfaces

yihyun-kim1 committed:
fix: Modify drag-and-drop with Object types

WIP)
- reflect the order in title when adding or deleting items
- problem with closing of the sheet after adding items

yihyun-kim1 committed:
fix: Modify drag-and-drop with Object types

WIP)
- reflect the order in title when adding or deleting items
- problem with closing of the sheet after adding items

yihyun-kim1 committed:
Merge branch 'main' into feature/#16

Brightbong92 commented on app/edit-course/page.tsx:
```diff
@@ -0,0 +1,221 @@
+"use client";
```
@yihyun-kim1 
app router방식에서 page.tsx는 SSR이므로 훅을 사용을 못하게 되었죠.
훅을 사용하기위한 코드들은 따로 _components 폴더를 만들어서 분리가 필요할것 같습니다.

yihyun-kim1 replied:
헉 제가 이 부분을 놓쳤네요 감사합니당 👍🏻

Brightbong92 commented on app/edit-course/page.tsx:
```diff
@@ -0,0 +1,221 @@
+
+  const handleItemDelete = (globalIndexToDelete: number) => {
+    setColumns((prevColumns) => {
+      const updatedList: Record<OrderType, ValueType[]> = {
+        food: [],
+        dessert: [],
+        beer: [],
+        play: [],
+      };
+
+      Object.values(prevColumns.course.list)
+        .flat()
+        .forEach((item) => {
+          if (item.globalIndex !== globalIndexToDelete) {
+            updatedList[item.type].push(item);
+          }
+        });
+
+      return {
+        ...prevColumns,
+        course: {
+          ...prevColumns.course,
+          list: updatedList,
+        },
+      };
```
setState안에 인라인 콜백 함수를 함수로 분리할수 있을것같은데 어떻게 생각하시나요?
prevcolumns, updatedList를 명시적 인자로 넘겨 계산한 데이터를 return하는 함수 일것같습니다.

yihyun-kim1 replied:
분리 후의 가독성이 훨씬 좋은 것 같아요 반영했습니다 👍🏻

Brightbong92 commented on app/edit-course/page.tsx:
```diff
@@ -0,0 +1,221 @@
+    Object.values(initialColumns.course.list).flat().length
+  );
+
+  const getAllItems = (): ValueType[] => {
+    return Object.values(columns.course.list)
+      .flat()
+      .sort((a, b) => a.globalIndex - b.globalIndex);
+  };
```
useMemo를 사용해도 괜찮을것 같은데 어떻게 생각하시나요?

yihyun-kim1 replied:
메모이제이션 사용 좋은 것 같아요! +_+ 한 번 고쳐보겠습니다

Brightbong92 commented on components/ui/sheet.tsx:
```diff
@@ -0,0 +1,145 @@
+      >
+        {children}
+        {/* 닫는 버튼은 사용하지 않아서 props 추가 */}
+        {showCloseButton && (
+          <SheetPrimitive.Close className="absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-secondary">
+            <X className="h-4 w-4" />
+            <span className="sr-only">Close</span>
+          </SheetPrimitive.Close>
+        )}
```
UI상 닫는버튼이 있을경우도 있으니 옵셔널로 추가해주신거 좋은것같습니다.

그러면 현재 바텀시트를 사용하는 컴포넌트들은 sheet.tsx 컴포넌트를 바로 사용하거나 기반으로 만들면 될것같아요.

본문에 논의사항으로 주신 바텀시트 생성관련해 저도 동의합니다.

yihyun-kim1 committed:
refactor: Separation of components and functions

- Change code that requires repetition to use useMemo hook
- Separate page-by-page components
